utensor_cgen/optimizer.py

```diff
@@ -0,0 +1,40 @@
+#-*- coding:utf8 -*-
+from collections import defaultdict
+
+
+class Optimizer(object):
+
+  @classmethod
+  def optimize(cls, ops_info, bfs_order, output_nodes, level='simple'):
+    """
+    [(op_name, op_info, ref_counts, to_eval), ...]
+    """
+    if level == 'simple':
+      return cls._simple_optimize(ops_info, bfs_order, output_nodes)
+    else:
+      raise ValueError("unknown optimization level: {}".format(level))
+
+  @classmethod
+  def _simple_optimize(cls, ops_info, bfs_order, output_nodes):
+    """Optimization using only reference count
+    """
+    optimized_order = []
+
+    refcnt_table = cls._tensor_ref_count(ops_info)
+    for op_name in bfs_order[::-1]:
+      op_info = ops_info[op_name]
+      if op_name in output_nodes or op_info.op_type in ["Const", "Placeholder"]:
+        to_eval = False
+      else:
+        to_eval = True
```

Comment:
So evaluate after every operator?

Reply:
Yes.
And let the reference counting do the job.

Reply:
cool, I talked this over with @neil-tan it makes sense